Enable encryption of files [Finishes #154378434]

habuma opened the pull request:
Note to reviewer: This is my first-ever significant contribution in the form of Go code. Be gentle.

glyn commented on docs/cli.md:
```diff
@@ -12,13 +12,15 @@ NAME:
 
-Note: if VALUE_TO_ENCRYPT contains characters that are special to the shell, it may be necessary
-to use escape characters or quotes. Refer to the shell's man page for details.
+      NOTE: VALUE_TO_ENCRYPT is optional if --file-to-encrypt flag is provided.
```
What does it mean if the value is provided in the file case?

I think it would be better to require that `VALUE_TO_ENCRYPT` must be omitted if the `--file-to-encrypt` flag is specified.

habuma replied:
It will ignore the value if `-f` is given. I documented that somewhere (just below the snippet you show here). I thought about making them mutually exclusive and failing if both are given, but quite honestly...it was getting late in the day and I'm on the road today (and will soon have spotty wifi), so I decided to go with what would get something checked in at the end of the day.

Wifi-permitting, I'll take another look and see about making them mutually exclusive.

glyn replied:
Thanks. It's always worth making user-facing interfaces as bullet-proof as possible.

glyn commented on config/encrypt.go:
```diff
@@ -10,23 +10,32 @@ import (
 }
 
-func EncryptWithResolver(cliConnection plugin.CliConnection, configServerInstanceName string, plainText string, authenticatedClient httpclient.AuthenticatedClient,
+func EncryptWithResolver(cliConnection plugin.CliConnection, configServerInstanceName string, plainText string, fileToEncrypt string, authenticatedClient httpclient.AuthenticatedClient,
```
Not sure about passing both `plainText` and `fileToEncrypt` when it would have been trivial to read the file in the caller and pass just `plainText` in either case.

habuma replied:
Yeah, I spent a lot of time thinking about that. Would you move the file-reading into `Encrypt`? I originally had it in `plugin.go` but felt that it really didn't belong there, so apparently I overcorrected.

glyn replied:
I think it's best to move it to `Encrypt`:
* you could, at least in principle, unit test this,
* it will avoid global clutter in `plugin.go`.

habuma replied:
I gave this a try on Friday and the hardest part was testing it...because I needed to send a fake resolver to `EncryptWithResolver` to test or else it would try to use a real resolver. The tests we have now call `EncryptWithResolver` directly, so this isn't a problem. But if I move the file loading logic up into `Encrypt`, then the tests would need to call `Encrypt` and I'd be faced with how to get a fake resolver in play.

Admittedly, by that time it was later in the day and I was not thinking as clearly, so I intend to give it another shot today. If you have any thoughts that might help, I'd be keen on hearing them.

glyn commented on config/encrypt.go:
```diff
@@ -10,23 +10,32 @@ import (
 		return "", err
 	}
 
+	var textToEncrypt string = plainText
```
The short form: `textToEncrypt := plainText` is more idiomatic.

glyn commented on config/encrypt_file_test.go:
```diff
@@ -0,0 +1,87 @@
+
+		Expect(fakeAuthClient.DoAuthenticatedPostCallCount()).Should(Equal(0))
+	})
+
```
I'd like to see tests for other cases of the file path, e.g. `../file.txt` pointing to a valid file, `/absolute/path/file.txt` pointing to a valid file, `somedir` pointing to a directory to make sure the diagnostics are suitable.

habuma replied:
Is there a way to mock the file reading? I ask because while these cases are interesting (and, in fact, I've confirmed that they work manually, because originally I was testing with a file in my home directory and I was giving an absolute path to it), I'm not sure how to test with a relative path so that the relative path will be available no matter where the tests are run (CI vs. local dev).

I suppose, however, I could add tests for `../somefile.txt` and for a `somedir`. It'd be cleaner, however, if I could mock the file loading so that I won't have to litter the project with test files. It felt awkward enough dropping the one test file alongside the Go code.

The only thing I can think of is to wrap the actual file-loading code with something that itself could be mocked. At that point, I'm not sure what it is that I'm testing...how I react to bogus files? I certainly wouldn't want write tests that only test the wrapper or the mock.

glyn replied:
I usually construct a suitable set of directories and files for the test in a temporary directory. Take a look at that kind of thing [here](https://github.com/cf-guardian/guardian/blob/80736895e96f5cf4051554ee71b71155cc795812/kernel/fileutils/fileutils_test.go) (unfortunately before I started using ginkgo).

With a bit in ingenuity, you can get all sorts of relative paths and edge cases.

glyn commented:
Putting the file reading logic in `Encrypt` might be counter-productive.
You could introduce another intermediate function that does the file
reading or else make the plainText parameter of EncryptWithResolver a
closure that returns a string and then pass in the relevant closure
depending on whether text or a file has been supplied. Separating the file
reading logic into a function would allow it to be separately tested.

On Mon, Jan 29, 2018 at 3:51 PM, Craig Walls <notifications@github.com>
wrote:

> *@habuma* commented on this pull request.
> ------------------------------
>
> In config/encrypt.go
> <https://github.com/pivotal-cf/spring-cloud-services-cli-plugin/pull/41#discussion_r164472136>
> :
>
> >  }
>
> -func EncryptWithResolver(cliConnection plugin.CliConnection, configServerInstanceName string, plainText string, authenticatedClient httpclient.AuthenticatedClient,
> +func EncryptWithResolver(cliConnection plugin.CliConnection, configServerInstanceName string, plainText string, fileToEncrypt string, authenticatedClient httpclient.AuthenticatedClient,
>
> I gave this a try on Friday and the hardest part was testing it...because
> I needed to send a fake resolver to EncryptWithResolver to test or else
> it would try to use a real resolver. The tests we have now call
> EncryptWithResolver directly, so this isn't a problem. But if I move the
> file loading logic up into Encrypt, then the tests would need to call
> Encrypt and I'd be faced with how to get a fake resolver in play.
>
> Admittedly, by that time it was later in the day and I was not thinking as
> clearly, so I intend to give it another shot today. If you have any
> thoughts that might help, I'd be keen on hearing them.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pivotal-cf/spring-cloud-services-cli-plugin/pull/41#discussion_r164472136>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFy6ZsMpzsMWooFBBmPZAr3HxhUYPgJks5tPejogaJpZM4Rtm2F>
> .
>



-- 
Regards,
Glyn

habuma commented:
@glyn : When you get a moment, can you review the latest, with the last couple of commits, to see what you think? I believe I've addressed most of your concerns, albeit perhaps not in the way you'd have done it.

At any rate, things are working as it is, so may I request that unless you have a significant concern (aka, this shouldn't be shipped as-is) then we go ahead and merge and chalk up any remaining minor concerns as potential improvement/refactor stories? I only ask this because this is a relatively high priority story and it's gone on a bit longer than planned.

glyn commented on docs/cli.md:
```diff
@@ -12,13 +12,15 @@ NAME:
    csev
+
+OPTIONS:
+   --f/--file-to-encrypt      A text file whose contents are to be encrypted. Cannot be used with VALUE_TO_ENCRYPT parameter.
```
This text is silent about the encoding that is expected for the text file. Ideally, I think this should be defined so that users know which encoding to use. Is it `UTF-8`, for example? The encoding is the one that is specified, or more likely taken as default, when the text is sent to the config server for encryption.

This [article](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/) is useful for anyone unfamiliar with encodings.

habuma replied:
It's my understanding (could be wrong) that Go principally deals with UTF-8 (with some limited support for UTF-16) and that working with any other encoding requires some additional packages.

Therefore, unless I'm terribly mistaken (a possibility), the encoding should be UTF-8. For documentation's sake, it's a reasonable qualification, and I've made such changes. If it turns out that other encodings are also suitable, then we can alter the documentation accordingly.

glyn replied:
Sounds good. Thanks.

glyn commented on test_support/file_operations.go:
```diff
@@ -0,0 +1,40 @@
+package test_support
```
The file name should be `file_operations_test.go` so that the code is only available to tests. (If it's used outside tests, then it's package would need renaming.)

habuma replied:
If I rename it to "file_operations_test.go", I get complaints that there are no non-test .go files in that directory. My options are: 1) name it as-is, 2) move it to the config package alongside the tests that use it, 3) put some dummy .go file alongside it to appease the compiler, 4) something else that I'm unaware of that will make it happy. Any thoughts?

glyn commented on plugin.go:
```diff
@@ -214,10 +220,10 @@ func (c *Plugin) GetMetadata() plugin.PluginMetadata {
 
-Note: if VALUE_TO_ENCRYPT contains characters that are special to the shell, it may be necessary
-to use escape characters or quotes. Refer to the shell's man page for details.`,
+      NOTE: Either VALUE_TO_ENCRYPT or --file-to-encrypt flag is required, but not both.`,
```
Words missing from the end of this sentence? E.g. "... but not both must be specified."

habuma replied:
No...that seems like a complete sentence, that describes what's required. I suppose that I could expand on it a little, but I was trying to remain brief since (speaking from my own experience reading such help info/docs) I wouldn't read anything longer anyway.

glyn replied:
I shouldn't quibble, but the English just seems wrong to me. How about "Either VALUE_TO_ENCRYPT or --file-to-encrypt flag must be provided, but not both."? If you can't see the improvement, let's agree to differ. :-)

glyn replied:
@fifthposition's suggestion is great as it avoids a passive construction.

glyn commented on docs/cli.md:
```diff
@@ -12,13 +12,15 @@ NAME:
 ALIAS:
    csev
+
+OPTIONS:
```
Has this file been regenerated with the latest help text? It doesn't seem to match what is in `plugin.go`.

habuma replied:
Done

glyn commented on config/encrypt_file_test.go:
```diff
@@ -0,0 +1,148 @@
+	})
+
+	JustBeforeEach(func() {
+		fakeAuthClient.DoAuthenticatedPostReturns(postResponse, postStatusCode, postErr)
```
This looks a bit odd. Usually, `JustBeforeEach` is used to drive the subject of the test, e.g. make the call to `Encrypt`. It can do more than that, but the idea is to avoid duplication. It can refer to variables which are set in `BeforeEach` blocks.

habuma replied:
This was "borrowed" in the copy-n-paste from encrypt_test.go. I kept it around because mainly I wasn't entirely sure what it was doing and also because if I removed it, then the tests would fail.

glyn replied:
Oh fair enough. Sorry for not realising that.

glyn commented on config/encrypt_file_test.go:
```diff
@@ -0,0 +1,148 @@
+		testFile := test_support.CreateFile(testDir, "file-to-encrypt.txt")
+		output, err = config.Encrypt(fakeCliConnection, serviceRegistryInstance, "", testFile, fakeAuthClient)
+
+		Expect(fakeAuthClient.DoAuthenticatedPostCallCount()).Should(Equal(1))
```
Testing the file processing as part of `Encrypt` means having to fake HTTP request handling. Did you explore the option of separating out a "middle layer" between `Encrypt` and `EncryptWithResolver` which could be tested separately while keeping the corresponding tests relatively focussed?

habuma replied:
I've made such an extraction with my latest commit. However...

I'd argue that there was already a middle layer in the `readFileContents()` function (albeit, contained within encrypt.go). That function is little more than a convenience wrapper around `ioutil.ReadFile()`, which converts the returned data to a string and prefixes the error text...little more than that. Any thorough testing I write around such an extraction would feel more like I'm just writing tests for `ioutil.ReadFile()`, which is obviously unnecessary.

Even if I did that, I'd feel compelled to write tests that flow through `Encrypt` for encrypting files...which is what I've already done. And, in fact, I've left the same (near-duplicated) tests in encrypt_file_test.go, because removing them would result in a loss of code coverage (at the little bit left in Encrypt that checks to make sure that there wasn't an error before attempting to encrypt the value).

habuma commented on config/encrypt.go:
```diff
@@ -43,3 +55,12 @@ func encrypt(plainText string, serviceURI string, accessToken string, authentica
+	if err != nil {
+		return "", fmt.Errorf("Error opening file at path %s : %s", fileToEncrypt, err)
+	}
+	var fileContents string = string(dat)
```
Done

habuma commented on config/encrypt_file_test.go:
```diff
@@ -0,0 +1,148 @@
+
+})
+
+// func check(err error) {
```
Done

fifthposition commented:
Maybe “Provide either VALUE_TO_ENCRYPT or the --file-to-encrypt flag, but not both.” ? :)

> On Jan 31, 2018, at 10:45 AM, Glyn Normington <notifications@github.com> wrote:
> 
> @glyn commented on this pull request.
> 
> In plugin.go:
> 
> >  
> -Note: if VALUE_TO_ENCRYPT contains characters that are special to the shell, it may be necessary
> -to use escape characters or quotes. Refer to the shell's man page for details.`,
> +      NOTE: Either VALUE_TO_ENCRYPT or --file-to-encrypt flag is required, but not both.`,
> I shouldn't quibble, but the English just seems wrong to me. How about "Either VALUE_TO_ENCRYPT or --file-to-encrypt flag must be provided, but not both."? If you can't see the improvement, let's agree to differ. :-)
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

glyn commented:
@habuma Looking great now, thanks. Just one maintainability improvement - please delete the relative path and non-existent directory tests from `encrypt_file_test.go` as these are exercised in `file_operations_test.go`. It's sufficient in `encrypt_file_test.go` to have the happy path and one file-related error path.

habuma commented:
@glyn : I've removed those tests. Will you now approve/merge?
